File: src/confy/dependency.zig
//:______________________________________________________________________
//  ᛝ confy  |  Copyright (C) Ivan Mar (sOkam!)  |  GNU LGPLv3 or later  :
//:______________________________________________________________________
//! @fileoverview Dependency Management
//______________________________________|
pub const Dependency = @This();
// @deps std
const std = @import("std");
// @deps zstd
const zstd = @import("../lib/zstd.zig");
const cstr = zstd.cstr;

name :cstr,
url  :cstr,
deps :?[]const Dependency= null,
opts :Dependency.Options= .{},

/// @descr Describes a sequence/list of {@link Dependency} objects
pub const List = zstd.seq(Dependency);

/// @descr Dependency Configuration Defaults
pub const default = struct {
  const src_subDir = "src";
};

pub const Options = struct{
  /// @descr Source code folder where the root file is stored
  src        :cstr=  Dependency.default.src_subDir,
  /// @descr
  ///  Path of the file that should be treated as the root/entry point of the dependency
  ///  Will use {@link Dependency.name}+".ext" when omitted.
  ///  Must be relative to {@link Options.src}.
  root       :?cstr=  null,
  /// @descr Whether the dependency should be treated as a git submodule or not.
  submodule  :bool=  false,
  /// @descr
  ///  When not null, override the target's default lib installation subpath with this folder
  ///  Must be relative to the current working directory _(`std.path.cwd()`)_ from where the builder is run.
  /// @note _Provided for completion. This option should almost never be needed._
  libDir     :?cstr=  null,
  /// @descr Subdependencies of this dependency. Only for construction, ignored everywhere else.
  deps       :?[]const Dependency= null,
}; //:: Dependency.Options

pub fn toNim (
    S : *const Dependency,
    D : cstr,
    A : std.mem.Allocator,
  ) !cstr {
  if (S.opts.src.len == 0) { return try std.fmt.allocPrint(A, "--path:{s}/{s}",     .{D, S.name});               }
  else                     { return try std.fmt.allocPrint(A, "--path:{s}/{s}/{s}", .{D, S.name, S.opts.src}); }
} //:: Dependency.toNim

const zig = struct {
  fn depOrModule (
      S : *const Dependency,
      D : cstr,
      M : bool,
      R : *zstd.str,
    ) !void {
    try R.appendSlice(if (M) " -M" else " --dep ");
    try R.appendSlice(S.name);
    if (!M) return;
    try R.append('=');
    try R.appendSlice(D);
    try R.append('/');
    try R.appendSlice(S.name);
    try R.append('/');
    try R.appendSlice(S.opts.src);
    try R.append('/');
    try R.appendSlice(S.opts.root orelse S.name);
    try R.appendSlice(".zig");
  }

  /// @descs Adds the given {@arg L} list of dependencies as `--dep name` at the end of the {@arg R} resulting string
  fn getDependencies (
      L : Dependency.List,
      R : *zstd.str,
    ) !void {
    if (L.items.len == 0) return;
    for (L.items) |dep| {
      try R.appendSlice(" --dep ");
      try R.appendSlice(dep.name);
    }
  }

  fn args (
      S    : *const Dependency,
      D    : cstr,
      root : bool,
      R    : *zstd.str,
    ) !void {
    for (S.deps orelse &.{}) |dep| std.debug.print("{s} ", .{dep.name});
    // Add the dependencies as --dep
    if (S.deps != null) for (S.deps.?) |dep| try zig.depOrModule(&dep, D, false, R);
    // Add the root at the end
    try zig.depOrModule(S, D, root, R);
  }
};
pub const toZig      = zig.args;
pub const getZigDeps = zig.getDependencies;


// TODO:
// pub fn clone(S :*const Dependency) void {
//   std.debug.print("...............\n", .{});
//   std.debug.print("Dependency.clone\n", .{});
//   std.debug.print("{s}\n", .{S.name});
//   std.debug.print("{s}\n", .{S.url});
//   std.debug.print("{s}\n", .{S.src});
//   std.debug.print("...............\n", .{});
// }

File: src/confy/dependency.nim
#:______________________________________________________________________
#  ᛝ confy  |  Copyright (C) Ivan Mar (sOkam!)  |  GNU GPLv3 or later  :
#:______________________________________________________________________
##! @fileoverview Tools for Dependency Management
#_________________________________________________|
# @deps std
from std/os import `/`
from std/strformat import `&`
from std/strutils import join
# @deps confy
import ./types/base
import ./types/errors
import ./types/build
import ./log

export build.Dependency
export build.Dependencies

# # Bash
# proc sh *(cmd :string; dbg :bool= false) :void=
#   if dbg: echo cmd
#   if cfg.fakeRun: return
#   if os.execShellCmd(cmd) != 0: raise newException(OSError, &"Failed to run shell command:  {cmd}")

func git *(
    trg  : build.BuildTarget;
    args : ArgsList;
  ) :void=
  ## @descr Runs git with the given {@arg args} using a shell.
  let cmd  = trg.cfg.git.bin&" "&args.join(" ")
  let code = os.execShellCmd( cmd )
  if code != 0: trg.fail GitError, &"Git returned `{code}` for command:\n  {cmd}"


#_______________________________________
# @section Dependency Auto-Download Management
#_____________________________
func download *(
    trg : build.BuildTarget;
    dep : build.Dependency;
  ) :void=
  for subdep in dep.deps: trg.download(subdep)
  var cmd :ArgsList= @[]
  if dep.submodule.active:
    var dir = dep.submodule.dir
    if dir == "" : dir = trg.cfg.dirs.src/"lib"/dep.name
    {.cast(noSideEffect).}:
      if os.dirExists(dir): return
    # Run   git submodule add
    cmd.add "submodule"
    cmd.add "add"
    cmd.add dep.url
    cmd.add dir
    trg.git cmd
    # Clear and run   git submodule update
    cmd = @[]
    cmd.add "submodule"
    cmd.add "update"
    cmd.add "--init"
    cmd.add "--recursive"
    cmd.add dir
  else:
    let dir = trg.cfg.dirs.lib/dep.name
    {.cast(noSideEffect).}:
      if os.dirExists(dir): return
    # Run   git clone
    cmd.add "clone"
    cmd.add "--recurse-submodules"
    cmd.add dep.url
    cmd.add dir
  trg.git cmd
#___________________
func download *(trg :build.BuildTarget; _:typedesc[Dependencies]) :void=
  for dep in trg.deps: trg.download(dep)


#_______________________________________
# @section Nim: Dependencies to Arguments
#_____________________________
func nim_path (
    dep  : build.Dependency;
    dir  : PathLike = ".";
  ) :build.Arg=
  result.add "--path:"
  result.add( dir/dep.name/dep.src )
#___________________
func toNim *(
    dep  : build.Dependency;
    dir  : PathLike = ".";
  ) :build.ArgsList= @[dep.nim_path(dir)]
#___________________
func toNim *(
    deps : build.Dependencies;
    dir  : PathLike = ".";
  ) :build.ArgsList=
  for dep in deps: result.add dep.toNim(dir)


#_______________________________________
# @section Zig: Dependencies to Arguments
#_____________________________
func depOrModule (
    dep  : build.Dependency;
    dir  : PathLike = ".";    ## Root Local Folder where the library should be found
    root : bool     = false;  ## Whether this is a root module or not
  ) :build.ArgsList=
  # Add as --dep name and return when not a root/entry module
  if not root: return @["--dep", dep.name]
  # Add as -Mname=path/to/file.zig when root
  let file =
    if dep.entry != "": dep.entry
    else              : dep.name&".zig"
  let path = dir/dep.name/dep.src/file
  result = @[&"-M{dep.name}={path}"]
#___________________
func toZig *(
    dep  : build.Dependency;
    dir  : PathLike = ".";    ## Root Local Folder where the library should be found
    root : bool = false;      ## Whether this is a root module or not
  ) :build.ArgsList=
  # Add the sub-dependencies as --dep
  for subdep in dep.deps: result &= subdep.depOrModule(dir, root=false)
  # Add the root at the end
  result.add dep.depOrModule(dir, root)
#___________________
func toZig *(deps :build.Dependencies) :build.ArgsList=
  if deps.len == 0: return @[]

